R/transformCounts.R

```diff
@@ -307,6 +346,109 @@ setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
+                               list(...)$scaling_constant, 1000)
+    
+    # Call .calc_scaling_factors method to calculate scaling factors
+    scaling_factors <- .calc_scaling_factors(mat, css_percentile)
+    
+    # Normalize the count data by dividing by the scaling factor
+    normalized_data <- sweep(mat, 2, scaling_factors / scaling_constant, "/")
+    return(normalized_data)
+}
+
+####################################.calc_css_fast##############################
+# This function applies cumulative sum scaling (CSS) to the abundance table 
+# using the fast variant method.
+.calc_css_fast <- function(mat, ...) {
+    # Check for the presence of the rel additional arg and set default if not present
+    rel <- ifelse("rel" %in% names(list(...)), list(...)$rel, 0.1)
+    
+    # Calculate the percentile using .calc_css_percentile
+    css_percentile <- .calc_css_percentile(mat, rel, ...)
+    
+    # Call .calc_css with the calculated percentile
+    normalized_data <- .calc_css(mat, css_percentile, ...)
+    return(normalized_data)
+}
+
+#################################.calc_css_percentile######################
+# Calculates the cumulative sum scaling (css) scaling percentiles from the given 
+# data
+#'  @importFrom matrixStats colSums2 colQuantiles rowMeans2 rowMedians
+.calc_css_percentile <- function(mat, rel, ...) {
+    # Replace zero values to NA, i.e. not detected
+    mat[ mat == 0 ] <- NA
+    # Calculate number of features detected
+    found_features <- colSums2(!is.na(mat))
+    # Stop if there are columsn with only 1 or 0 found features
+    if( any(found_features <= 1) ){
+        stop(
+            "There are samples that contain 1 or less found features.",
+            "'fast CSS' method cannot be applied.", call. = FALSE)
+    }
+    # Calculate quantiles. Take into account only positive, found features
+    quantiles <- colQuantiles(
+        mat, prob = seq(0, 1, length.out = max(found_features)), na.rm = TRUE)
+    quantiles <- t(quantiles)
+    # Sort the columns based on abundance
+    mat <- apply(mat, 2, function(col){
+        col[ is.na(col) ] <- 0
+        col <- sort(col)
+        return(col)
+    })
+    # Compute the row means of sorted matrix to create a reference profile
+    ref <- rowMeans2(mat)
+    ref <- ref[ ref > 0 ]
+    # Calculate the differences between the reference profile and each
+    # column's quantiles
+    difference <- ref - quantiles
+    # Compute the row medians of absolute differences
+    difference <- rowMedians(abs(difference))
+    # Find the first point where the relative change exceeds the threshold
+    res <- abs(diff(difference)) / difference[-1]
+    res <- which(res > rel)
+    res <- res[[1]] / max(found_features)
+    # If the calculated percentile is less than or equal to 0.50, use 0.50 as
+    # the default value
+    if( res < 0.50 ){
+        res <- 0.50
+        message("Default value being used (", res, ").")
+    }
+    return(res)
+}
```

Comment:
I simplified the code and removed loops. The functionality should be the same. Check that it is still working as you expect